Utils/azuretools-core/src/com/microsoft/azuretools/telemetrywrapper/DefaultProducer.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    @Override
+    public void endTransaction(String eventName, String operName, Map<String, String> properties, long time) {
+        try {
+            Map<String, Double> metrics = new HashMap<>();
+            metrics.put(DURATION, Double.valueOf(time));
+            Map<String, String> mergedProperty = mergeProperties(addOperNameAndId(properties, operName));
+            if (errorInfo.get() != null) {
+                mergedProperty.putAll(errorInfo.get());
+            }
+            sendTelemetry(EventType.opEnd, eventName, mergedProperty, metrics);
+        } catch (Exception ignore) {
+        } finally {
+            errorInfo.remove();
+            operIDS.remove();
+        }
+    }
+
+    @Override
+    public void sendError(String eventName, String operName, ErrorType errorType, String errMsg,
+        Map<String, String> properties) {
+        try {
+            Map<String, String> errorMap = new HashMap<>();
+            errorMap.put(ERROR_CODE, "1");
+            errorMap.put(ERROR_MSG, errMsg);
+            errorMap.put(ERROR_TYPE, errorType.name());
+            // we need to save errorinfo, and then write the error info when we end the transaction, by this way we
+            // can quickly get the operation result from opend
+            errorInfo.remove();
+            errorInfo.set(errorMap);
+
+            Map<String, String> newProperties = addOperNameAndId(properties, operName);
+            newProperties.putAll(errorMap);
+            sendTelemetry(EventType.error, eventName, mergeProperties(newProperties), null);
+        } catch (Exception ignore) {
+        }
+    }
+
+    @Override
+    public void sendInfo(String eventName, String operName, Map<String, String> properties) {
+        try {
+            sendTelemetry(EventType.info, eventName, mergeProperties(addOperNameAndId(properties, operName)), null);
+        } catch (Exception ignore) {
+        }
+    }
+
+    @Override
+    public void sendWarn(String eventName, String operName, Map<String, String> properties) {
+        try {
+            sendTelemetry(EventType.warn, eventName, mergeProperties(addOperNameAndId(properties, operName)), null);
+        } catch (Exception ignore) {
+        }
+    }
+
+    @Override
+    public void setTelemetryClient(TelemetryClient client) {
+        this.client = client;
+    }
+
+    private Map<String, String> addOperNameAndId(Map<String, String> properties, String operName) {
+        Map<String, String> result = new HashMap<>();
+        if (properties != null) {
+            result.putAll(properties);
+        }
+        result.put(OPERATION_NAME, operName);
+        String operId = operIDS.get();
+        result.put(OPERATION_ID, operId == null ? UUID.randomUUID().toString() : operId);
+        return result;
+    }
+
+    private Map<String, String> mergeProperties(Map<String, String> properties) {
+        Map<String, String> commonProperties = TelemetryManager.getInstance().getCommonProperties();
```

Comment:
what about
``` java
Map<String, String> merged = new HashMap<>(commonProperties);
if(properties!=null){
    merged.putAll(properties)
}
```

or we could return a new hashmap when calls 'TelemetryManager.getInstance().getCommonProperties()' as invoker could modify common properties if they get the original object, which is dangerous.

Reply:
commonproperties is a unmodified map